sklearn/preprocessing/knn_imputation.py

```diff
@@ -0,0 +1,253 @@
+# Authors: Ashim Bhattarai <ashimb9@gmail.com>
+# License: BSD 3 clause
+
+from __future__ import division
+import warnings
+import numpy as np
+
+from ..base import BaseEstimator, TransformerMixin
+from ..utils import check_array
+from ..utils.validation import check_is_fitted
+from ..utils.validation import FLOAT_DTYPES
+from ..neighbors import NearestNeighbors
+from ..neighbors.base import _get_weights, _check_weights
+__all__ = [
+    'KNNImputer',
+]
+
+
+def _get_mask(X, value_to_mask):
+    """Compute the boolean mask X == missing_values."""
+    if value_to_mask == "NaN" or np.isnan(value_to_mask):
+        return np.isnan(X)
+    else:
+        return X == value_to_mask
+
+
+class KNNImputer(BaseEstimator, TransformerMixin):
+    """Imputation transformer for completing missing values using Nearest
+    Neighbors. Broadly speaking, the imputation is performed using either
+    the weighted or the unweighted mean of the desired number of neighbors.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default="NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    n_neighbors : int, optional (default = 5)
+        Number of neighbors to get.
+
+    weights : str or callable
+        weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights.  All points in each neighborhood
+          are weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - [callable] : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
+        Uniform weights are used by default.
+
+    metric : string or callable, optional (default = 'masked_euclidean')
+        metric to use for distance computation.
+
+    row_max_missing: float, optional (default = 0.5)
```

Comment:
We've only recently put in those checks, but they're useful. I'd show you how this is currently rendered, but it's not being rendered in the docs because you've not yet added it to `doc/modules/classes.rst`. Please do.